### PR TITLE
Sanitize `HOST_PROMPT_SUPPORT` notifications

### PR DESCRIPTION
**Pull request by The-EG**
### Description

Currently, when `HOST_PROMPT_SUPPORT` is enabled, many of the status messages displayed on the LCD are echoed back as 'action:notification' commands. This isn't ideal for two reasons:
 1. it's not consistent. For example, `M190` will send a notification of 'Bed heating...', but `M109` does not send a corresponding 'E1 heating...' notification.
 2. **status messages are not notifications!** Not every `M117` message needs to be echoed back to the host. A good example, having a 'Layer N of T' message at the begining of each layer in your slicer config, or using an OctoPrint Plugin like DisplayLayerProgress will cause a notification for every single `M117`.

Number 1 is due to the fact that `MarlinUI` has 3 (that I can find) methods for setting the status message. 2 of these send a host notification when `HOST_PROMPT_SUPPORT` are enabled. As you can probably guess by now, `M109` uses the third method 😄.

For number 2: I suspect these notifications are being abused to show status messages on hosts that are functioning as a printer LCD screen. Ie. TFTs or even full blown hosts like OctoDash running on top of OctoPrint. While this works, doing this on a 'real' host, like OctoPrint itself or pronterface, etc doesn't make sense. The host knows the contents of the `M117` command before it is sent, because its the one sending it...there is no need to echo it back!

Outside of the exception for a host that is abusing the messages to show status messages, notifications should be just that: messages that are interesting enough that the host should get the user's attention, maybe even by showing a popup. They should be fairly rare.

This adds the host notification to the 3rd method mentioned above and adds a configuration option: `HOST_STATUS_NOTIFICATIONS`. If enabled, any time the UI status message is updated using any of the 3 functions mentioned above, it will also be echoed back as an `action:notifiication`. Without that enabled, only specific conditions trigger a notification.

Currently:
 - EEPROM Settings Load/Save/Reset
 - PID tune success/error
 - where ever it was already explicitly being sent without relying on one of the UI set status methods to do it. I didn't find any of these, but there may be a few hiding out there...

### Requirements

`HOST_ACTION_COMMANDS` and `HOST_PROMPT_SUPPORT`

### Benefits

Printing from a host will no longer be bombarded by every single status message without the new configuration option. The few notifications that remain are worthy of attention.

### Configurations

As tested on my Ender 3 v2:
[ender 3 v2.zip](https://github.com/MarlinFirmware/Marlin/files/7228424/ender.3.v2.zip)

But, any configuration with `HOST_ACTION_COMMANDS` and `HOST_PROMPT_SUPPORT` enabled should suffice.

### Related Issues

N/A

### CODE REVIEW

**Comment by The-EG** [comment]
Hmm I'll look into that failed test. I may need some `#if` guards around the added notifications in a few places. Is there a recommended example config I can use for the 'rambo' test?

**Comment by The-EG** [comment]
> Is there a recommended example config I can use for the 'rambo' test?

Nevermind. It is possible to run the CI tests locally on Windows if you setup an msys2 environment...

File: Marlin/src/module/settings.cpp [hunk 1]
```diff
@@ -1471,7 +1475,10 @@ void MarlinSettings::postprocess() {
         store_mesh(ubl.storage_slot);
     #endif
 
-    if (!eeprom_error) LCD_MESSAGEPGM(MSG_SETTINGS_STORED);
+    if (!eeprom_error) {
+      LCD_MESSAGEPGM(MSG_SETTINGS_STORED);
+      TERN_(HOST_PROMPT_SUPPORT, host_action_notify_P(GET_TEXT(MSG_SETTINGS_STORED)));
```
Comment (thinkyhead) [inline Marlin/src/module/settings.cpp:1480]: `HOST_PROMPT_SUPPORT` or `HOST_STATUS_NOTIFICATIONS` ?

File: Marlin/src/module/settings.cpp [hunk 2]
```diff
@@ -2363,12 +2371,14 @@ void MarlinSettings::postprocess() {
...
         DEBUG_ERROR_MSG("EEPROM CRC mismatch - (stored) ", stored_crc, " != ", working_crc, " (calculated)!");
         TERN_(DWIN_CREALITY_LCD_ENHANCED, ui.set_status(GET_TEXT(MSG_ERR_EEPROM_CRC)));
+        TERN_(HOST_PROMPT_SUPPORT, host_action_notify_P(GET_TEXT(MSG_ERR_EEPROM_CRC)));
         IF_DISABLED(EEPROM_AUTO_INIT, ui.eeprom_alert_crc());
       }
       else if (!validating) {
         DEBUG_ECHO_START();
         DEBUG_ECHO(version);
         DEBUG_ECHOLNPGM(" stored settings retrieved (", eeprom_index - (EEPROM_OFFSET), " bytes; crc ", (uint32_t)working_crc, ")");
+        TERN_(HOST_PROMPT_SUPPORT, host_action_notify("Stored settings retrieved"));
```
Comment (thinkyhead) [inline Marlin/src/module/settings.cpp]: The above output is only emitted when debugging.

File: Marlin/src/module/settings.cpp [hunk 3]
```diff
@@ -3028,6 +3038,7 @@ void MarlinSettings::reset() {
   postprocess();
 
   DEBUG_ECHO_MSG("Hardcoded Default Settings Loaded");
+  TERN_(HOST_PROMPT_SUPPORT, host_action_notify("Hardcoded Default Settings Loaded"));
```
Comment (thinkyhead) [inline Marlin/src/module/settings.cpp]: Debug only.

File: Marlin/src/module/temperature.cpp [hunk 4]
```diff
@@ -747,12 +749,14 @@ volatile bool Temperature::raw_temps_ready = false;
...
         TERN_(DWIN_CREALITY_LCD_ENHANCED, DWIN_PidTuning(PID_TUNING_TIMEOUT));
         TERN_(EXTENSIBLE_UI, ExtUI::onPidTuning(ExtUI::result_t::PID_TUNING_TIMEOUT));
+        TERN_(HOST_PROMPT_SUPPORT, host_action_notify(STR_PID_TIMEOUT));
         SERIAL_ECHOLNPGM(STR_PID_TIMEOUT);
         break;
       }
 
       if (cycles > ncycles && cycles > 2) {
         SERIAL_ECHOLNPGM(STR_PID_AUTOTUNE_FINISHED);
+        TERN_(HOST_PROMPT_SUPPORT, host_action_notify("PID Autotune Finished!")); // TODO: New str value?
```
Comment (thinkyhead) [inline Marlin/src/module/temperature.cpp]: Text sent to the host in this manner should be translatable. Only serial messages associated with host protocols should be hardcoded in English.

**Comment by The-EG** [comment]
I'm seeing the 'debug' ones in the serial output without having enabled anything specific for it, but I'll remove those just the same. 

The main thing is: with these changes, without `HOST_STATUS_NOTIFICATIONS` you won't get any notifications because all of the existing ones rely on setting the status message. I was trying to at least provide a few to start with, but in any case, meaningful notifications can be added as they are desired later.

File: Marlin/src/lcd/marlinui.cpp [hunk 5]
```diff
@@ -1348,7 +1348,7 @@ constexpr uint8_t epps = ENCODER_PULSES_PER_STEP;
   void MarlinUI::set_status(const char * const message, const bool persist) {
     if (alert_level) return;
 
-    TERN_(HOST_PROMPT_SUPPORT, host_action_notify(message));
+    TERN_(HOST_STATUS_NOTIFICATIONS, host_action_notify(message));
```
Comment (thinkyhead) [inline Marlin/src/lcd/marlinui.cpp]: One caveat with this concept is that the status message may be changed very often, either by G-code or by the host itself or one of its plugins. Note that the "Machine Ready" message, the current printing file's name, or some other non-null string is used as a response to "resetting" the status message.

**Comment by thinkyhead** [comment]
> I'm seeing the 'debug' ones in the serial output without having enabled anything specific for it,

The debug serial macros are activated in response to…
```cpp
#define DEBUG_OUT EITHER(EEPROM_CHITCHAT, DEBUG_LEVELING_FEATURE)
```

**Comment by InsanityAutomation** [comment]
To sum up some of the discussion on discord, In my opinion this loses functionality that has a negligible cost to maintain with little to no apparent gain, as well as sets a path that will have notable labor investment to flesh out and maintain. Right now there is only a few messages here. Next people will start adding other notifications as they deem them important. When enough have had the added code block to override inserted, it'll be too much again and split to yet another config option causing even more inconsistency in the output. The growing slope of complexity around it is quite large.

While "abuse" by hosts to cause notifications is possible, as octoprint plugins and slicer scripts can do it, my believe is the machine should do as its told. There may be just as valid use cases for the plugin to provide a prompt through the machine to again show a display in both consistent locations of what is needed. As, the primary goal is to get the notifications from commands such as G34 iterations, G29 probe status, PID completion, and other machine generated status output as well as relevant fault displays to a host this blocks all required messaging in its current state with only a few explicitly defined messages actually sent.

The intention was for not only headless machines where octoprint is the only UI (See Makergear for a commercially available example) but to also maintain consistency so that you can see exactly whats going on at the machine without having to physically go there. Anything that causes these to deviate in my opinion is an issue. Treat the external display status the same as an internal display. I see no difference to an internal notify in the status bar compared to an external notify with the action command.

In order to maintain the critical portion while disabling the minor notifications, one would have to go through and add additional output code to every single desired message, cluttering the code and creating yet another maintenance headache. Another option that maintains consistency may be to add the ability to define persistent messages which M117 cannot override, and incorporate either a timeout period or specific events to clear these.

Another option discussed with an Octo plugin dev who may chime in themselves was potential for a terminal style filter on the notification display. Then the host side can control what notifications they do and do not want to display. This makes the most sense in my opinion.

**Comment by jneilliii** [comment]
Hello, I'm the other Octo plugin dev @InsanityAutomation alludes to. I definitely think that having all these messages echoed by default is beneficial, for the case in point you've already mentioned, OctoDash for example. It currently has a PR that allows for it to display action notifications for a similar setup as Insanity already mentioned where there is no other LCD attached. The PR also came from a printer manufacturer in the process of delivering a unit with that OctoPrint/OctoDash combination. 

I personally see both points, but from the "what's important to me" perspective, it would make sense for the host software, in my case OctoPrint, to be able to filter those messages the way I see fit, and not controlled solely by the firmware. Give me all the data and I'll decide what I want to show and not show. Since I don't contribute to Marlin, I have no input on the maintainability of the code, but could see if you are managing the filter on the firmware side it could get hard to deal with because of that "what's important to me" perspective again.

**Comment by The-EG** [comment]
What I don't understand about the push back here is that I'm not asking anyone to use this feature. Have it disabled by default, so that the default behavior doesn't change from what it is currently. 

But, please, *please*, do not force every single user of Marlin to use your use case. That is absurd.

**Comment by InsanityAutomation** [comment]
@The-EG its the disable all notifications EXCEPT that causes the biggest concern. It could be defensible if it was based on classification (such as all warn / error only) even though I still see those as much better handled filtered by the host. However as it sits it creates a slope of what fits into this notification method. Code for each notification someone deems important... Cold extrude, Probe fail, Heater timeout... One by one things will get added until its seen as too much again. Its a long term headache to maintain. I don't see a cause for fragmentation of expected output from the machine side, especially if there is a willingness to add filtering to the host side.

**Comment by The-EG** [comment]
I'm going to sum this up by: I still disagree and I think this is an improvement that could lead to eventual improvements in the future. 

Ultimately, if those with the ability to merge this don't agree, that is fine.

@InsanityAutomation I also understand being called out directly by name in Discord chat could be frustrating, however I do want to point out that *I* was not that person. While I think we can agree to disagree here, if you are trying to convince me of something because of that earlier conversation, you are speaking to the wrong person.

**Comment by InsanityAutomation** [comment]
@The-EG the insinuation of a hit and run code drop of some kind from me being the breakage when it was likely the recent serial macro changes, not code that's been stable for ages is frustrating and will be addressed with that person later. Its still a separate item to this. If you were filtering based on only sending WARN or ERROR, I may not like it much but wouldn't really argue it this much. Needing to explicitly set individual messages everywhere however I do see as a problem long term.

**Comment by thinkyhead** [comment]
I've just been tweaking the host action code to make it into a singleton class and looking at adding an enable / disable mechanism with a set of flags for different types of notifications and prompts. Serial controllers could use host prompts if they included handling, but AFAIK right now only some host software — namely OctoPrint — is implementing this notification and prompt scheme. So for now I just put in one set of flags (which are currently set with `M111 H`) and pertain to all serial ports. I haven't added filters to any prompts or notifications yet. I'll just start by adding the capability to turn all notifications on/off, and later I'll look at filtering different types, such as status messages, errors, etc.

**Pull request merged**